### PR TITLE
Fix bincount for empty input

### LINKED ISSUE #9756: torch.bincount errors on empty input
## Issue description

`torch.bincount` gives an error when given an empty tensor, rather than just returning zero counts like `numpy.bincount`.  Trips up edge cases when counting filtered data.

## Code example

```
>>> import torch, numpy
>>> numpy.bincount([1]), numpy.bincount([]), numpy.bincount([], minlength=10)
(array([0, 1]), array([], dtype=int64), array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0]))

>>> torch.bincount(torch.LongTensor([]))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: bincount only supports 1-d non-negative integral inputs.

>>> torch.bincount(torch.LongTensor([]), minlength=10)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: bincount only supports 1-d non-negative integral inputs.
```

### PR DESCRIPTION
**Pull request by vishwakftw**
Added tests too. Fixes #9756 .

### CODE REVIEW

**Comment by zou3519** [comment]
I'm a little concerned because the documentation is ill-defined for what happens in this case: https://pytorch.org/docs/master/torch.html?highlight=bincount#torch.bincount . If we're trying to match numpy, we should probably make the docs clearer on what happens in this edge case.

**Comment by vishwakftw** [comment]
Sure, I will clear up that as well.

**Comment by vishwakftw** [comment]
@zou3519 Are we not allowed to create 0-element tensors anymore? It seems `torch.tensor([], ...)` is failing.

**Comment by vishwakftw** [comment]
It was an nit in the ordering of the `if` clauses. Sorry about that.

**Comment by vishwakftw** [comment]
Is this good to go?

**Comment by vishwakftw** [comment]
@ailzhang Is this good to go?

File: torch/_torch_docs.py [hunk 1]
```diff
@@ -609,14 +609,17 @@ def parse_kwargs(desc):
 :math:`out[n] += weights[i]` if :attr:`weights` is specified else
 :math:`out[n] += 1`.
 
+If the number of elements in :attr:`input` is 0, then the result is a tensor of size
```
Comment (gchanan) [inline torch/_torch_docs.py]: what if `minlength` isn't specified?  The documentation says it's optional.

File: torch/_torch_docs.py [hunk 2]
```diff
@@ -609,14 +609,17 @@ def parse_kwargs(desc):
...
+
 Arguments:
     input (Tensor): 1-d int tensor
     weights (Tensor): optional, weight for each value in the input tensor.
         Should be of same size as input tensor.
     minlength (int): optional, min number of bins. Should be non-negative.
 
 Shape:
-    output (Tensor): ``Size([max(input) + 1])``
+    output (Tensor): ``Size([max(input) + 1])`` if :attr:`input` is non-empty.
```
Comment (gchanan) [inline torch/_torch_docs.py]: would be good to put the shape of the output here if `input` is empty.

File: torch/_torch_docs.py (hunk 1 above)
Comment (gchanan) [reply torch/_torch_docs.py]: Why don't we just add it to the above, e.g.

```
The number of bins (size 1) is one larger than the largest value in :attr:`input`, unless :attr:`input` is empty, in which case the result is size 0.`

**Comment by vishwakftw** [comment]
@SsnL Is this good to go?

**Review by soumith — APPROVED** [review]

**Comment by soumith** [comment]
the merge got botched for this one. working on re-merging!

**Pull request closed without merging**